src/common/application.cpp

```diff
@@ -273,24 +278,63 @@ void Application::requestExit()
 {
     gIsRunning = false;
     io_context_.stop();
+    if (engine_)
+    {
+        // Let the engine perform shutdown tasks, if any.
+        engine_->onShutdown();
```

Comment:
Could this not go into the engine destructor, and then there doesn't need to be the conditional logic here?